### ISSUE TITLE
Inference not working as expected

### ISSUE THREAD

**Issue by sledorze** [Microsoft/TypeScript#8772]
**TypeScript Version:** 1.9.0-dev.20160523-1.0

**Code**

```ts
// A self-contained demonstration of the problem follows...

class Foo {

  identity<T>(x: T): T {
    return x
  }

  compose<A, B, C>(f: (b: B) => C, g: (a: A) => B) : (a:A) => C {
    return (a: A) => f(g(a))
  }

  constructor()  {
    let value = this.compose((a : number) => a, this.identity)
  }
}

```

**Expected behavior:**

value should have type (a: number) => number
OR
it should issue an error as the tsconfig.json has "noImplicitAny" defined to true.

**Actual behavior:** 

value has type (a: {}) => number
AND
no error from the compiler.

**Comment by mhegazy** [comment]
duplicate of https://github.com/Microsoft/TypeScript/issues/5254

**Comment by sledorze** [comment]
@mhegazy why has it been closed? I see that the duplicate issue is still open..

**Comment by mhegazy** [comment]
> why has it been closed? I see that the duplicate issue is still open..

We like to keep on report to track an issue. no need to have multiple open reports. Duplicate reports are closed with a link to the main report.

**Comment by sledorze** [comment]
thanks @mhegazy for the clarification, I understand your need to reduce the 'noise' :)
